xbattbar.man
.\"
.\" $Id: xbattbar.man,v 1.13.2.1 2001/02/02 05:25:29 suguru Exp $
.\"
.\" Copyright (c) 1998-2001 Suguru Yamaguchi <suguru@wide.ad.jp>
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published
.\" by the Free Software Foundation; either version 2 of the License, or (at
.\" your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.Dd December 5, 1998
.Dt XBATTBAR 1
.Os
.Sh NAME
.Nm xbattbar
.Nd show battery status in X Window
.Sh SYNOPSIS
.Nm xbattbar
.Op Fl a 
.Op Fl t Ar thickness
.Op Fl p Ar interval
.Op Fl I Ar color
.Op Fl O Ar color
.Op Fl i Ar color
.Op Fl o Ar color
.Op Ar top | bottom | left | right
.Sh DESCRIPTION
.Nm xbattbar
shows the current (laptop) battery status in the X window environment.
Battery status is obtained through the APM kernel module.
The APM kernel module depends on your UNIX platform,
therefore, consult your documentation for its detail.
.Pp
.Nm xbattbar
shows its battery status in a simple bar indicator.
.Nm -a
option makes the indicator window keep always on top of your screen.
The thickness of the indicator is 3 pixels in default and
you can set the thickness as a parameter of 
.Nm -t
option.
This indicator is appeared in the bottom of the display, as its default.
The option
.Nm top,
.Nm left,
or
.Nm right
forces the status indicator to be at the top, left, and right of the
display,
respectively.
Though it's a default, you can also use
.Nm bottom
as the option.
.Pp
In the case the AC line is on-line (plugged in),
its color of the bar indicator consists of "green" and "olive drab"
portions.
The "green" portion shows its battery charging level.
These colors can be changed by
.Nm -I
and
.Nm -O
options.
In the other case (AC line is off-line),
the "blue" portion of the bar indicator shows the percentage of
its remaining battery level.
The color conbination for the AC off-line case can be changed by
.Nm -i
and
.Nm -o
options.
.Pp
.Nm xbattbar
trys to know its battery status in every 10 seconds in default.
This is achived by APM polling.
.Nm -p
option sets the polling interval in second.
.Pp
If the mouse cursor enters in the status indicator,
the diagnosis window appears in the center of the display,
which shows both AC line status and battery remaining level.
This diagnosis window disappears if the mouse cursor leaves from
the status indicator.
.Sh SEE ALSO
.Xr xbatt 1
\- an official battery status check command on BSD/OS 3.0,
(but it's not fancy).
.Sh AUTHOR
Suguru Yamaguchi <suguru@wide.ad.jp>,
Akira Kato <kato@wide.ad.jp>,
Noriyuki Shigechika <nazo@wide.ad.jp>


xbattbar.c
/* -*- mode: c; tab-width: 2; c-basic-offset: 2; indent-tabs-mode: nil -*-
 * vim: set ts=2 sw=2 expandtab:
 *
 * Copyright (c) 2025 Toby Slight <tslight@pm.me>
 * Copyright (c) 1998-2001 Suguru Yamaguchi <suguru@wide.ad.jp>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

static char *ReleaseVersion="1.5.0";

#include <err.h>
#include <sys/types.h>
#include <sys/time.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/file.h>
#include <sys/ioctl.h>
#include <X11/Xlib.h>
#include <X11/Xft/Xft.h>

#define PollingInterval 30	/* APM polling interval in sec */
#define BI_THICKNESS    3	/* battery indicator thickness in pixels */

#define BI_Bottom	0
#define BI_Top		1
#define BI_Left		2
#define BI_Right	3
#define BI_Horizontal	((bi_direction & 2) == 0)
#define BI_Vertical	((bi_direction & 2) == 2)

#define myEventMask (ExposureMask|EnterWindowMask|LeaveWindowMask|VisibilityChangeMask)
#define DiagXMergin 20
#define DiagYMergin 5
/* default font & time format of dialog window */
#define DiagFont "monospace:bold:size=18"
#define TimeFormat "%H:%M %A %d %B %Y"
#define WarningLevel 10
#define CriticalLevel 5

/*
 * Global variables
 */
int ac_line = -1;               /* AC line status */
int battery_level = -1;         /* battery level */

/* for battery remaining estimation */
int elapsed_time = 0;
char remainbuf[32];
size_t remainbuf_size = sizeof(remainbuf);

int alwaysontop = True;

struct itimerval IntervalTimer;     /* APM polling interval timer */

int bi_direction = BI_Left;         /* status bar location */
int bi_height;                      /* height of Battery Indicator */
int bi_width;                       /* width of Battery Indicator */
int bi_x;                           /* x coordinate of upper left corner */
int bi_y;                           /* y coordinate of upper left corner */
int bi_thick = BI_THICKNESS;        /* thickness of Battery Indicator */
int bi_interval = PollingInterval;  /* interval of polling APM */

Display *disp;
int screen, dispwidth, dispheight;
Colormap colormap;
unsigned long black, green, magenta, olive, blue, red;
Window winbar;                  /* bar indicator window */
Window winstat = -1;            /* battery status window */
GC gcbar;
GC gcstat;
unsigned int width,height;
XEvent theEvent;

/*
 * function prototypes
 */
void InitDisplay(void);
void battery_check(void);
void plug_proc(int);
void battery_proc(int);
void redraw(void);
void showdiagbox(void);
void disposediagbox(void);
long getcolor(const char *);
void usage(char **);
void about_this_program(void);
void estimate_remain(void);

/*
 * usage of this command
 */
void about_this_program(void)
{
  fprintf(stderr,
	  "This is xbattbar version %s, "
	  "copyright (c)1998-2001 Suguru Yamaguchi\n",
	  ReleaseVersion);
}

void usage(char **argv)
{
  fprintf(stderr,
	  "\n"
	  "usage:\t%s [-a] [-h|v] [-p sec] [-t thickness]\n"
	  "\t\t[ top | bottom | left | right ]\n"
	  "-a:     always on top. [def: True].\n"
	  "-v, -h: show this message.\n"
	  "-t:     bar (indicator) thickness. [def: 3 pixels]\n"
	  "-p:     polling interval. [def: 10 sec.]\n"
	  "top, bottom, left, right: bar localtion. [def: \"left\"]\n",
	  argv[0]);
  exit(0);
}

/*
 * InitDisplay:
 * create small window in top or bottom
 */
void InitDisplay(void)
{
  Window root;
  int x,y;
  unsigned int border,depth;
  XSetWindowAttributes att;

  if((disp = XOpenDisplay(NULL)) == NULL) {
    fprintf(stderr, "xbattbar: can't open display.\n");
    exit(1);
  }

  screen     = DefaultScreen(disp);
  dispwidth  = DisplayWidth(disp, screen);
  dispheight = DisplayHeight(disp, screen);
  colormap   = DefaultColormap(disp, DefaultScreen(disp));
  black	     = getcolor("black");
  magenta    = getcolor("magenta");
  green	     = getcolor("green");
  olive	     = getcolor("olive drab");
  blue	     = getcolor("blue");
  red        = getcolor("red");

  if(XGetGeometry(disp, DefaultRootWindow(disp), &root, &x, &y,
		  &width, &height, &border, &depth) == 0) {
    fprintf(stderr, "xbattbar: can't get window geometry\n");
    exit(1);
  }

  switch (bi_direction) {
  case BI_Top: /* (0,0) - (width, bi_thick) */
    bi_width = width;
    bi_height = bi_thick;
    bi_x = 0;
    bi_y = 0;
    break;
  case BI_Bottom:
    bi_width = width;
    bi_height = bi_thick;
    bi_x = 0;
    bi_y = height - bi_thick;
    break;
  case BI_Left:
    bi_width = bi_thick;
    bi_height = height;
    bi_x = 0;
    bi_y = 0;
    break;
  case BI_Right:
    bi_width = bi_thick;
    bi_height = height;
    bi_x = width - bi_thick;
    bi_y = 0;
  }

  winbar = XCreateSimpleWindow(disp, DefaultRootWindow(disp),
			       bi_x, bi_y, bi_width, bi_height,
			       0, BlackPixel(disp,0), WhitePixel(disp,0));

  /* make this window without its titlebar */
  att.override_redirect = True;
  XChangeWindowAttributes(disp, winbar, CWOverrideRedirect, &att);

  XMapWindow(disp, winbar);

  gcbar = XCreateGC(disp, winbar, 0, 0);
}

int main(int argc, char **argv)
{
  extern char *optarg;
  extern int optind;
  int ch;

  about_this_program();
  while ((ch = getopt(argc, argv, "at:f:h:p:v")) != -1)
    switch (ch) {
    case 'a':
      alwaysontop = True;
      break;
    case 't':
    case 'f':
      bi_thick = atoi(optarg);
      break;
    case 'p':
      bi_interval = atoi(optarg);
      break;
    case 'h':
    case 'v':
      usage(argv);
      break;
    }
  argc -= optind;
  argv += optind;

  if (argc > 0) {
    if (strcasecmp(*argv, "top") == 0)
      bi_direction = BI_Top;
    else if (strcasecmp(*argv, "bottom") == 0)
      bi_direction = BI_Bottom;
    else if (strcasecmp(*argv, "left") == 0)
      bi_direction = BI_Left;
    else if (strcasecmp(*argv, "right") == 0)
      bi_direction = BI_Right;
  }

  /*
   * set APM polling interval timer
   */
  IntervalTimer.it_interval.tv_sec = (long)bi_interval;
  IntervalTimer.it_interval.tv_usec = (long)0;
  IntervalTimer.it_value.tv_sec = (long)1;
  IntervalTimer.it_value.tv_usec = (long)0;
  if ( setitimer(ITIMER_REAL, &IntervalTimer, NULL) != 0 ) {
    fprintf(stderr,"xbattbar: can't set interval timer\n");
    exit(1);
  }

  /*
   * X Window main loop
   */
  InitDisplay();
  signal(SIGALRM, (void *)(battery_check));
  battery_check();
  XSelectInput(disp, winbar, myEventMask);
  while (1) {
    XWindowEvent(disp, winbar, myEventMask, &theEvent);
    switch (theEvent.type) {
    case Expose:
      /* we redraw our window since our window has been exposed. */
      redraw();
      break;

    case EnterNotify:
      /* create battery status message */
      showdiagbox();
      break;

    case LeaveNotify:
      /* destroy status window */
      disposediagbox();
      break;

    case VisibilityNotify:
      if (alwaysontop) XRaiseWindow(disp, winbar);
      break;

    default:
      /* for debugging */
      fprintf(stderr,
	      "xbattbar: unknown event (%d) captured\n",
	      theEvent.type);
    }
  }
}

void redraw(void)
{
  if (ac_line) {
    plug_proc(battery_level);
  } else {
    battery_proc(battery_level);
    if (battery_level < WarningLevel) showdiagbox();
  }
}

long getcolor(const char *color)
{
  int rc;
  XColor tcolor;
  if (!(rc = XAllocNamedColor(disp, colormap, color, &tcolor, &tcolor)))
    errx(1, "can't allocate %s", color);

  return tcolor.pixel;
}

void showdiagbox(void)
{
  XSetWindowAttributes att;
  int boxw, boxh;
  char diagmsg[64];
  XftDraw *xftdraw = NULL;
  XftFont *xftfont;
  XftColor xftcolor;
  XGlyphInfo extents;

  estimate_remain();

  sprintf(diagmsg, "%s: %d%%%s", ac_line ? "Charging" : "Discharging",
	  battery_level, remainbuf);
  xftfont = XftFontOpenName(disp, screen, DiagFont);
  if (!xftfont)
    errx(1, "XftFontOpenName failed for %s", DiagFont);
  // Get width and height of message
  XftTextExtentsUtf8(disp, xftfont, (FcChar8 *)diagmsg, strlen(diagmsg),
		     &extents);

  boxw = extents.width + 20;
  boxh = extents.height + 20;

  if(winstat != -1) disposediagbox();
  winstat = XCreateSimpleWindow(disp, DefaultRootWindow(disp),
				(dispwidth-boxw)/2,
				(dispheight-boxh)/2,
				boxw, boxh,
				1,       // width
				magenta, //border
				black);  // background

  att.override_redirect = True;
  XChangeWindowAttributes(disp, winstat, CWOverrideRedirect, &att);
  XMapWindow(disp, winstat);

  xftdraw = XftDrawCreate(disp,
			  winstat,
			  DefaultVisual(disp, screen),
			  colormap);
  XRenderColor render_color = { 0x0000,   // red
				0xffff,   // green
				0x0000,   // blue
				0xffff }; // opacity
  XftColorAllocValue(disp, DefaultVisual(disp, screen),
		     colormap, &render_color, &xftcolor);
  XftDrawStringUtf8(xftdraw, &xftcolor, xftfont, 10, 30, (FcChar8 *)diagmsg,
		    strlen(diagmsg));

  // Free Xft resources
  XftDrawDestroy(xftdraw);
  XftFontClose(disp, xftfont);
  XftColorFree(disp, DefaultVisual(disp, screen),
	       colormap, &xftcolor);
}

void disposediagbox(void)
{
  if ( winstat != -1 ) {
    XDestroyWindow(disp, winstat);
    winstat = -1;
  }
}

void battery_proc(int left)
{
  int pos;
  if (BI_Horizontal) {
    pos = width * left / 100;
    XSetForeground(disp, gcbar, blue);
    XFillRectangle(disp, winbar, gcbar, 0, 0, pos, bi_thick);
    XSetForeground(disp, gcbar, red);
    XFillRectangle(disp, winbar, gcbar, pos, 0, width, bi_thick);
  } else {
    pos = height * left / 100;
    XSetForeground(disp, gcbar, blue);
    XFillRectangle(disp, winbar, gcbar, 0, height-pos, bi_thick, height);
    XSetForeground(disp, gcbar, red);
    XFillRectangle(disp, winbar, gcbar, 0, 0, bi_thick, height-pos);
  }
  XFlush(disp);
}

void plug_proc(int left)
{
  int pos;
  if (BI_Horizontal) {
    pos = width * left / 100;
    XSetForeground(disp, gcbar, green);
    XFillRectangle(disp, winbar, gcbar, 0, 0, pos, bi_thick);
    XSetForeground(disp, gcbar, olive);
    XFillRectangle(disp, winbar, gcbar, pos+1, 0, width, bi_thick);
  } else {
    pos = height * left / 100;
    XSetForeground(disp, gcbar, green);
    XFillRectangle(disp, winbar, gcbar, 0, height-pos, bi_thick, height);
    XSetForeground(disp, gcbar, olive);
    XFillRectangle(disp, winbar, gcbar, 0, 0, bi_thick, height-pos);
  }
  XFlush(disp);
}

void estimate_remain(void)
{
  if (battery_level > 98 && ac_line) {
    remainbuf[0] = '\0';
    return;
  }

  static int battery_base = -1;
  int diff, remain;

  /* static value initialize */
  if (battery_base == -1) {
    battery_base = battery_level;
    remainbuf[0] = '\0';
    return;
  }

  diff = battery_base - battery_level;

  if (diff == 0) return;

  /* estimated time until battery critical */
  if (diff > 0) {
    remain = elapsed_time * (battery_level - CriticalLevel) / diff ;
    remain = remain * bi_interval;  /* in sec */
    if (remain < 0 ) remain = 0;
    snprintf(remainbuf, remainbuf_size, ": %02d:%02d:%02d -> Critical",
	     remain / 3600, (remain % 3600) / 60, remain % 60);
    elapsed_time = 0;
    battery_base = battery_level;
    return;
  }

  /* estimated time until battery charged */
  remain = elapsed_time * (battery_level - 100) / diff;
  remain = remain * bi_interval;  /* in sec */
  snprintf(remainbuf, remainbuf_size, ": %02d:%02d:%02d -> Charged",
	   remain / 3600, (remain % 3600) / 60, remain % 60);
  elapsed_time = 0;
  battery_base = battery_level;
}



#ifdef __bsdi__

#include <machine/apm.h>
#include <machine/apmioctl.h>

int first = 1;
void battery_check(void)
{
  int fd;
  struct apmreq ar ;

  ++elapsed_time;

  ar.func = APM_GET_POWER_STATUS ;
  ar.dev = APM_DEV_ALL ;

  if ((fd = open(_PATH_DEVAPM, O_RDONLY)) < 0) {
    perror(_PATH_DEVAPM) ;
    exit(1) ;
  }
  if (ioctl(fd, PIOCAPMREQ, &ar) < 0) {
    fprintf(stderr, "xbattbar: PIOCAPMREQ: APM_GET_POWER_STATUS error 0x%x\n", ar.err);
  }
  close (fd);

  if (first || ac_line != ((ar.bret >> 8) & 0xff)
      || battery_level != (ar.cret&0xff)) {
    first = 0;
    ac_line = (ar.bret >> 8) & 0xff;
    battery_level = ar.cret&0xff;
    redraw();
  }
  signal(SIGALRM, (void *)(battery_check));
}

#endif /* __bsdi__ */

#ifdef __FreeBSD__

#include <sys/types.h>
#include <sys/sysctl.h>

#if defined(__i386__)
#include <machine/apm_bios.h>

#define APMDEV21       "/dev/apm0"
#define APMDEV22       "/dev/apm"

#define        APM_STAT_UNKNOWN        255

#define        APM_STAT_LINE_OFF       0
#define        APM_STAT_LINE_ON        1

#define        APM_STAT_BATT_HIGH      0
#define        APM_STAT_BATT_LOW       1
#define        APM_STAT_BATT_CRITICAL  2
#define        APM_STAT_BATT_CHARGING  3
#endif /* i386 */

int first = 1;
void battery_check(void)
{
  int r, p;
  static int sysctl = 1;

  if (sysctl) {
    size_t r_size, p_size;

    /* get current status */
    r_size = sizeof(r);
    if (sysctl && sysctlbyname("hw.acpi.battery.life", &r, &r_size, NULL, 0) == -1) {
#if defined(__i386__)
      fprintf(stderr, "xbattbar: fall back to apm interface\n");
      sysctl = 0;
#else /* !i386 */
      fprintf(stderr, "xbattbar: can not get battery status\n");
      exit(1);
#endif /* i386 */
    }

    /* get AC-line status */
    p_size = sizeof(p);
    if (sysctl && sysctlbyname("hw.acpi.acline", &p, &p_size, NULL, 0) == -1) {
#if defined(__i386__)
      fprintf(stderr, "xbattbar: fall back to apm interface\n");
      sysctl = 0;
#else /* !i386 */
      fprintf(stderr, "xbattbar: can not get AC-line status\n");
      exit(1);
#endif /* i386 */
    }
  }

#if defined(__i386__)
  if (!sysctl) {
    int fd;
    struct apm_info info;

    if ((fd = open(APMDEV21, O_RDONLY)) == -1 &&
	(fd = open(APMDEV22, O_RDONLY)) == -1) {
      fprintf(stderr, "xbattbar: cannot open apm device\n");
      exit(1);
    }
    if (ioctl(fd, APMIO_GETINFO, &info) == -1) {
      fprintf(stderr, "xbattbar: ioctl APMIO_GETINFO failed\n");
      exit(1);
    }
    close (fd);

    /* get current status */
    if (info.ai_batt_life == APM_STAT_UNKNOWN) {
      switch (info.ai_batt_stat) {
      case APM_STAT_BATT_HIGH:
	r = 100;
	break;
      case APM_STAT_BATT_LOW:
	r = 40;
	break;
      case APM_STAT_BATT_CRITICAL:
	r = 10;
	break;
      default:        /* expected to be APM_STAT_UNKNOWN */
	r = 100;
      }
    } else if (info.ai_batt_life > 100) {
      /* some APM BIOSes return values slightly > 100 */
      r = 100;
    } else {
      r = info.ai_batt_life;
    }

    /* get AC-line status */
    if (info.ai_acline == APM_STAT_LINE_ON) {
      p = APM_STAT_LINE_ON;
    } else {
      p = APM_STAT_LINE_OFF;
    }
  }
#endif /* i386 */

  ++elapsed_time;

  if (first || ac_line != p || battery_level != r) {
    first = 0;
    ac_line = p;
    battery_level = r;
    redraw();
  }
  signal(SIGALRM, (void *)(battery_check));
}

#endif /* __FreeBSD__ */

#ifdef __NetBSD__

#include <machine/apmvar.h>

#define _PATH_APM_SOCKET       "/var/run/apmdev"
#define _PATH_APM_CTLDEV       "/dev/apmctl"
#define _PATH_APM_NORMAL       "/dev/apm"

int first = 1;
void battery_check(void)
{
  int fd, r, p;
  struct apm_power_info info;

  if ((fd = open(_PATH_APM_NORMAL, O_RDONLY)) == -1) {
    fprintf(stderr, "xbattbar: cannot open apm device\n");
    exit(1);
  }

  if (ioctl(fd, APM_IOC_GETPOWER, &info) != 0) {
    fprintf(stderr, "xbattbar: ioctl APM_IOC_GETPOWER failed\n");
    exit(1);
  }

  close(fd);

  ++elapsed_time;

  /* get current remoain */
  if (info.battery_life > 100) {
    /* some APM BIOSes return values slightly > 100 */
    r = 100;
  } else {
    r = info.battery_life;
  }

  /* get AC-line status */
  if (info.ac_state == APM_AC_ON) {
    p = APM_AC_ON;
  } else {
    p = APM_AC_OFF;
  }

  if (first || ac_line != p || battery_level != r) {
    first = 0;
    ac_line = p;
    battery_level = r;
    redraw();
  }
}

#endif /* __NetBSD__ */


#ifdef linux

#include <errno.h>
#include <linux/apm_bios.h>

#define APM_PROC	"/proc/apm"

#define        APM_STAT_LINE_OFF       0
#define        APM_STAT_LINE_ON        1

typedef struct apm_info {
  const char driver_version[10];
  int        apm_version_major;
  int        apm_version_minor;
  int        apm_flags;
  int        ac_line_status;
  int        battery_status;
  int        battery_flags;
  int        battery_percentage;
  int        battery_time;
  int        using_minutes;
} apm_info;


int first = 1;
void battery_check(void)
{
  int r,p;
  FILE *pt;
  struct apm_info i;
  char buf[100];

  /* get current status */
  errno = 0;
  if ( (pt = fopen( APM_PROC, "r" )) == NULL) {
    fprintf(stderr, "xbattbar: Can't read proc info: %s\n", strerror(errno));
    signal(SIGALRM, (void *)(battery_check));
    exit(1);
  }

  fgets( buf, sizeof( buf ) - 1, pt );
  buf[ sizeof( buf ) - 1 ] = '\0';
  sscanf( buf, "%s %d.%d %x %x %x %x %d%% %d %d\n",
	 &i.driver_version,
	 &i.apm_version_major,
	 &i.apm_version_minor,
	 &i.apm_flags,
	 &i.ac_line_status,
	 &i.battery_status,
	 &i.battery_flags,
	 &i.battery_percentage,
	 &i.battery_time,
	 &i.using_minutes );

  fclose (pt);

  ++elapsed_time;

  /* some APM BIOSes return values slightly > 100 */
  if ( (r = i.battery_percentage) > 100 ){
    r = 100;
  }

  /* get AC-line status */
  if ( i.ac_line_status == APM_STAT_LINE_ON) {
    p = APM_STAT_LINE_ON;
  } else {
    p = APM_STAT_LINE_OFF;
  }

  if (first || ac_line != p || battery_level != r) {
    first = 0;
    ac_line = p;
    battery_level = r;
    redraw();
  }
  signal(SIGALRM, (void *)(battery_check));
}

#endif /* linux */
